### PR TITLE
Fix pods stuck in Pending state forever

### LINKED ISSUE #49314: Pod that failed to bind, stuck in Pending state forever
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

Pod got stuck in Pending state forever when failed to bind to host. 

**What you expected to happen**:

For it to be rescheduled on another host

**How to reproduce it (as minimally and precisely as possible)**:

Not always reproducible. But simply created an RC replicas=2 on a 3 host setup, and one of the Pods got stuck in pending state for a while. Following error messages were found in the scheduler logs:

```
7/20/2017 8:30:40 AME0720 15:30:40.435703       1 scheduler.go:282] Internal error binding pod: (scheduler cache ForgetPod failed: pod test-677550-rc-edit-namespace/nginx-jvn09 state was assumed on a different node)
7/20/2017 8:31:11 AMW0720 15:31:11.119885       1 cache.go:371] Pod test-677550-rc-edit-namespace/nginx-jvn09 expired
```

Describe pod output:

```
nginx-jvn09   0/1       Pending   0          2h
> kubectl describe pod nginx-jvn09 --namespace=test-677550-rc-edit-namespace
Name:           nginx-jvn09
Namespace:      test-677550-rc-edit-namespace
Node:           /
Labels:         name=nginx
Annotations:    kubernetes.io/created-by={"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"Replicatio
nController","namespace":"test-677550-rc-edit-namespace","name":"nginx","uid":"6313354d-6d60-11e...
Status:         Pending
IP:
Controllers:    ReplicationController/nginx
Containers:
  nginx:
    Image:              nginx
    Port:               80/TCP
    Environment:        <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-vz8x6 (ro)
Volumes:
  default-token-vz8x6:
    Type:       Secret (a volume populated by a Secret)
    SecretName: default-token-vz8x6
    Optional:   false
QoS Class:      BestEffort
Node-Selectors: <none>
Tolerations:    node.alpha.kubernetes.io/notReady=:Exists:NoExecute for 300s
                node.alpha.kubernetes.io/unreachable=:Exists:NoExecute for 300s
Events:         <none>
```

RC yml:

```
apiVersion: v1
kind: ReplicationController
metadata:
  name: nginx
spec:
  replicas: 2
  selector:
    name: nginx
  template:
    metadata:
      labels:
        name: nginx
    spec:
      containers:
        - name: nginx
          image: nginx
          ports:
            - containerPort: 80
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.7.1
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release): RHEL7, Docker native 1.12.6
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by julia-stripe**
**What this PR does / why we need it**:

This PR fixes 2 bugs introduced into the scheduler in https://github.com/kubernetes/kubernetes/commit/ecb962e6585#diff-67f2b61521299ca8d8687b0933bbfb19R223.

Together, they meant that when a pod fails to bind to a node, it can end up stuck in the Pending state indefinitely. The way this happened was:

1. The pod fails to bind to the node
2. `ForgetPod` fails, because the wrong `*v1.Pod` struct was passed into `bind` and therefore `ForgetPod`  (`pod` instead of `assumed`) -- `ForgetPod` requires that the pod ask it to forget matches the pod in the cache
3. `bind` returns immediately after ForgetPod fails instead of correctly reporting an error, because of this line: https://github.com/kubernetes/kubernetes/commit/ecb962e6585#diff-67f2b61521299ca8d8687b0933bbfb19R223 -- this should be a `glog.Errorf`, not a return. The return on that line causes the function to skip the error handling, which means that the pod isn't put back onto the queue to attempt scheduling again.
4. Because the pod isn't put back on to the queue, the pod stays stuck in `Pending` forever until the scheduler is restarted.

This also adds better error handling when `AssumePod` fails -- it seems to me that any time pod scheduling fails, we should handle the error with `sched.config.Error` and put the pod back in the queue to be rescheduled instead of silently failing and leaving the pod potentially stuck in `Pending` forever.

It also adds 2 new unit tests, which test:

- when both `ForgetPod` and `Bind` fail, the correct error is reported
- when `AssumePod` fails, the correct error is reported
- when `Bind` fails, the correct pod is passed into ForgetPod

**Which issue this PR fixes** 
fixes #49314 

**Special notes for your reviewer**:

**Release note**:

```release-note
Fix bug where pods stuck in Pending state forever after binding fails
```

### CODE REVIEW

File: plugin/pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -222,7 +225,7 @@ func (sched *Scheduler) bind(assumed *v1.Pod, b *v1.Binding) error {
 	if err != nil {
```
Comment (k82cn) [inline plugin/pkg/scheduler/scheduler.go]: we should call FinishBinding just after bind as before.

File: plugin/pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -185,7 +185,7 @@ func (sched *Scheduler) schedule(pod *v1.Pod) (string, error) {
 }
 
 // assume signals to the cache that a pod is already in the cache, so that binding can be asnychronous.
-func (sched *Scheduler) assume(pod *v1.Pod, host string) error {
+func (sched *Scheduler) assume(pod *v1.Pod, host string) (*v1.Pod, error) {
```
Comment (k82cn) [inline plugin/pkg/scheduler/scheduler.go]: we should not return assumed pod, we will check current Pod with the pod in cache.

**Comment by k82cn** [comment]
/assign

Good catch!, add some comments :).

File: plugin/pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -194,13 +194,16 @@ func (sched *Scheduler) assume(pod *v1.Pod, host string) error {
...
-		// as binding doesn't make sense anyway.
-		// This should be fixed properly though.
-		return err
+
+		sched.config.Error(pod, err)
+		sched.config.Recorder.Eventf(pod, v1.EventTypeWarning, "FailedScheduling", "AssumePod failed: %v", err)
+		sched.config.PodConditionUpdater.Update(pod, &v1.PodCondition{
+			Type:    v1.PodScheduled,
+			Status:  v1.ConditionFalse,
+			Reason:  "AssumePodFailed",
```
Comment (k82cn) [inline plugin/pkg/scheduler/scheduler.go]: `Assume` is internal phase of scheduler, we should not publish it; different scheduler may have different way to handle race condition between cache & apiserver.

File: plugin/pkg/scheduler/scheduler.go (hunk 1 above)
Comment (julia-stripe) [inline plugin/pkg/scheduler/scheduler.go]: the call to `FinishBinding` was moved in this PR: https://github.com/kubernetes/kubernetes/pull/48451 -- let me know if you think it should be moved back (and why? :) ).

File: plugin/pkg/scheduler/scheduler.go (hunk 2 above)
Comment (julia-stripe) [inline plugin/pkg/scheduler/scheduler.go]: sorry, I don't understand -- the reason this is returning the assumed pod is that `bind` expects to get the assumed pod as an argument (`bind(assumed *v1.Pod, b *v1.Binding)`). 

before the refactor assume and bind were part of the same function and the assumed pod got passed to `ForgetPod` and `FinishBinding` -- you can see the old code (which I believe was correct)  here: https://github.com/kubernetes/kubernetes/blob/80f26fa8a89ef5863cb19c71a620bb389d025166/plugin/pkg/scheduler/scheduler.go#L229-L234

File: plugin/pkg/scheduler/scheduler.go (hunk 3 above)
Comment (julia-stripe) [inline plugin/pkg/scheduler/scheduler.go]: I'm not sure what the best way to indicate "this is most likely a bug in the scheduler" is but changed it to `SchedulerError` .

File: plugin/pkg/scheduler/scheduler.go (hunk 1 above)
Comment (k82cn) [inline plugin/pkg/scheduler/scheduler.go]: finishbinding will add deadline and remove the resumed pod when expired.

Comment (julia-stripe) [inline plugin/pkg/scheduler/scheduler.go]: do you think we should revert #48451?

File: plugin/pkg/scheduler/scheduler.go (hunk 2 above)
Comment (k82cn) [inline plugin/pkg/scheduler/scheduler.go]: wow, yes. We did not use object from cache directly; prefer to highlight copy outside `assume`:

```
assumedPod := *pod

sched.assume(assumedPod)
go sched.bind(assumedPod)
```

File: plugin/pkg/scheduler/scheduler.go (hunk 1 above)
Comment (k82cn) [inline plugin/pkg/scheduler/scheduler.go]: Give me some time to check history.

That's error-prone part, we used to close some PRs which want to enhance it. Another concern is that our UT and e2e did not catch it :(.

File: plugin/pkg/scheduler/scheduler.go (hunk 2 above)
Comment (julia-stripe) [inline plugin/pkg/scheduler/scheduler.go]: Done, thanks!

**Comment by cblecker** [comment]
@julia-stripe This is currently failing verify as `plugin/pkg/scheduler/testing/fake_cache.go` needs a gofmt :)

**Comment by julia-stripe** [comment]
oops, fixed!

can someone add the `cherrypick-candidate` label and release: 1.7.0? (is that the right way to flag bug fixes for the 1.7 release?)

**Comment by cblecker** [comment]
@julia-stripe One of the scheduling maintainers should be able to assist with that once:
- This PR is approved to merge into master
- It's determined that this bug appears in 1.7, and the fix is unlikely to cause any regressions.

More information on the cherry-pick procedure is available [here](https://git.k8s.io/community/contributors/devel/cherry-picks.md). This would fall under the section "INDIVIDUAL CHERRYPICKS".

File: plugin/pkg/scheduler/scheduler.go (hunk 1 above)
Comment (k82cn) [inline plugin/pkg/scheduler/scheduler.go]: yup, still prefer to call FinishBinding just after bind as before. But we can not revert #48451 , we need to just log error if 'FinishBinding' failed.

Current fix did not get a chance to call 'FinishBinding' which will still leave assumed pod in cache, so the pod can not be re-scheduled. For the concern on race of #48451, I think that's fine; 'ForgetPod' is quick enough, and scheduler will correct it in next cycle if it happened.

Comment (k82cn) [inline plugin/pkg/scheduler/scheduler.go]: /cc @wojtek-t @timothysc  for more suggestion :).

Comment (julia-stripe) [inline plugin/pkg/scheduler/scheduler.go]: Updated this PR to call `FinishBinding` right after `bind` and log an error instead of returning. (thank you so much for all the comments!)

Comment (julia-stripe) [inline plugin/pkg/scheduler/scheduler.go]: Also happy to break that commit out as a separate PR since it's an independent problem & this is getting kind of big, not sure which is better

**Comment by julia-stripe** [comment]
PTAL @k82cn 

Fixed all the issues you mentioned! Is there anything else? Should https://github.com/kubernetes/kubernetes/pull/49661/commits/589cb36c92e5e1897437ff1a89d8cd6bc0a76f23 be broken out into a separate PR because it's an independent change?

**Comment by julia-stripe** [comment]
Also pulled one smaller change out of this in the hopes that it's easier to review: https://github.com/kubernetes/kubernetes/pull/50028

File: plugin/pkg/scheduler/scheduler.go (hunk 1 above)
Comment (k82cn) [inline plugin/pkg/scheduler/scheduler.go]: As we had created another PRs, can we simplify this PR to only include `FinishBinding` and `ForgetPod`?

**Comment by julia-stripe** [comment]
Going to close this in favour of #50028 and open a separate PR to improve the error handling in scheduler.go.

**Comment by julia-stripe** [comment]
@k82cn Opened https://github.com/kubernetes/kubernetes/pull/50106 which is the error handling parts of this PR -- hopefully that will be easier to review.

**Comment by k82cn** [comment]
great, thanks very much :).
